modulemd/v2/modulemd-defaults.c

```diff
@@ -0,0 +1,309 @@
+
+  klass = MODULEMD_DEFAULTS_GET_CLASS (self);
+  g_return_val_if_fail (klass->get_mdversion, 0);
+
+  return klass->get_mdversion (self);
+}
+
+
+static void
+modulemd_defaults_set_module_name (ModulemdDefaults *self,
+                                   const gchar *module_name)
+{
+  g_return_if_fail (MODULEMD_IS_DEFAULTS (self));
+
+  /* It is a coding error if we ever get a NULL name here */
+  g_return_if_fail (module_name);
+
+  /* It is a coding error if we ever get the default name here */
+  g_return_if_fail (g_strcmp0 (module_name, DEF_DEFAULT_NAME_STRING));
+
+  ModulemdDefaultsPrivate *priv =
+    modulemd_defaults_get_instance_private (self);
+
+  g_clear_pointer (&priv->module_name, g_free);
+  priv->module_name = g_strdup (module_name);
+
+  g_object_notify_by_pspec (G_OBJECT (self), properties[PROP_MODULE_NAME]);
+}
+
+
+const gchar *
+modulemd_defaults_get_module_name (ModulemdDefaults *self)
+{
+  ModulemdDefaultsPrivate *priv =
+    modulemd_defaults_get_instance_private (self);
+
+  return priv->module_name;
+}
+
+
+static void
+modulemd_defaults_get_property (GObject *object,
+                                guint prop_id,
+                                GValue *value,
+                                GParamSpec *pspec)
+{
+  ModulemdDefaults *self = MODULEMD_DEFAULTS (object);
+
+  switch (prop_id)
+    {
+    case PROP_MODULE_NAME:
+      g_value_set_string (value, modulemd_defaults_get_module_name (self));
+      break;
+
+    case PROP_MDVERSION:
+      g_value_set_uint64 (value, modulemd_defaults_get_mdversion (self));
+      break;
+
+    default: G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+modulemd_defaults_set_property (GObject *object,
+                                guint prop_id,
+                                const GValue *value,
+                                GParamSpec *pspec)
+{
+  ModulemdDefaults *self = MODULEMD_DEFAULTS (object);
+
+  switch (prop_id)
+    {
+    case PROP_MODULE_NAME:
+      modulemd_defaults_set_module_name (self, g_value_get_string (value));
+      break;
+
+    case PROP_MDVERSION:
+    default: G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+modulemd_defaults_class_init (ModulemdDefaultsClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = modulemd_defaults_finalize;
+  object_class->get_property = modulemd_defaults_get_property;
+  object_class->set_property = modulemd_defaults_set_property;
+
+  klass->copy = modulemd_defaults_default_copy;
+  klass->validate = modulemd_defaults_default_validate;
+
+  properties[PROP_MDVERSION] = g_param_spec_uint64 (
+    "mdversion",
+    "Metadata Version",
+    "The metadata version of this Defaults object. Defaults to the "
+    "highest-supported version if not specified at object construction. "
+    "Cannot be changed after object construction except through the upgrade() "
```

Comment:
Note that the `upgrade()` routine doesn't modify this on the object it receives, but creates a new one.
Thta makes this phrasing sound incorrect to me.

Reply:
You're right. I'll rephrase.